proxy/coco/src/request/waiting_room.rs

```diff
@@ -0,0 +1,700 @@
+    pub fn insert<R>(&mut self, urn: RadUrn, request: R)
+    where
+        R: Into<SomeRequest<T>>,
+    {
+        self.requests.insert(urn, request.into());
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::error;
+
+    use librad::{keys::SecretKey, peer::PeerId, uri::RadUrn};
+    use pretty_assertions::assert_eq;
+    use proptest::{collection, prelude::prop_assert_eq};
+
+    use super::*;
+
+    #[test]
+    fn happy_path_of_full_request() -> Result<(), Box<dyn error::Error + 'static>> {
+        let mut waiting_room: WaitingRoom<()> = WaitingRoom::new(Config::default());
+        let urn: RadUrn = "rad:git:hwd1yre85ddm5ruz4kgqppdtdgqgqr4wjy3fmskgebhpzwcxshei7d4ouwe"
+            .parse()
+            .expect("failed to parse the urn");
+        let peer = PeerId::from(SecretKey::new());
+        let request = waiting_room.create(urn.clone(), ());
+
+        assert_eq!(request, None);
+
+        let strategy: Strategy<rand::rngs::StdRng> = Strategy::First;
+        let created = waiting_room.next(strategy);
+        assert_eq!(created, Some(&Request::new(urn.clone(), ())),);
+
+        waiting_room.queried(&urn, ())?;
+        let expected = SomeRequest::Requested(Request::new(urn.clone(), ()).request(()));
+        assert_eq!(waiting_room.get(&urn), Some(&expected));
+
+        waiting_room.found(&urn, peer.clone(), ())?;
+        let expected = SomeRequest::Found(
+            Request::new(urn.clone(), ())
+                .request(())
+                .first_peer(peer.clone(), ()),
+        );
+        assert_eq!(waiting_room.get(&urn), Some(&expected));
+
+        waiting_room.cloning(&urn, peer.clone(), ())?;
+        let expected = SomeRequest::Cloning(
+            Request::new(urn.clone(), ())
+                .request(())
+                .first_peer(peer.clone(), ())
+                .cloning(MAX_QUERIES, MAX_CLONES, peer.clone(), ())
+                .unwrap_right(),
+        );
+        assert_eq!(waiting_room.get(&urn), Some(&expected));
+
+        let found_repo = urn.clone();
+
+        waiting_room.cloned(&urn, found_repo.clone(), ())?;
+        let expected = SomeRequest::Cloned(
+            Request::new(urn.clone(), ())
+                .request(())
+                .first_peer(peer.clone(), ())
+                .cloning(MAX_QUERIES, MAX_CLONES, peer, ())
+                .unwrap_right()
+                .cloned(found_repo, ()),
+        );
+        assert_eq!(waiting_room.get(&urn), Some(&expected));
+
+        Ok(())
+    }
+
+    #[test]
+    fn cannot_create_twice() {
+        let mut waiting_room: WaitingRoom<()> = WaitingRoom::new(Config::default());
+        let urn: RadUrn = "rad:git:hwd1yre85ddm5ruz4kgqppdtdgqgqr4wjy3fmskgebhpzwcxshei7d4ouwe"
+            .parse()
+            .expect("failed to parse the urn");
+        waiting_room.create(urn.clone(), ());
+        let request = waiting_room.create(urn.clone(), ());
+
+        assert_eq!(request, Some(SomeRequest::Created(Request::new(urn, ()))));
```

Comment:
Doesn't this only prove that `request` was "successful" in creating the request?

Reply:
The returned request _is_ the original request. If it was a fresh request then it would have been `None`. I can add to the test though to move it into the next state and show we get back that request

Reply:
That would make it clearer!